File: MonotonicStack/496_Next_Greater_Element_I/496_Next_Greater_Element_I.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Stack<Integer> stk = new Stack<>();
        stk.push(Integer.MAX_VALUE);
        HashMap<Integer,Integer> hash = new HashMap<>();
        List<Integer> res = new ArrayList<>();

        for( int num: nums2 ){
            while( num > stk.peek() ) {
                Integer key = stk.pop();
                hash.put(key, num);
            }
            stk.push(num);
        }  
        
        for( int num: nums1 ){
            if(hash.containsKey(num)) {
                res.add(hash.get(num));
            } else {
                res.add(-1);
            }
        }
        int[] result = new int[res.size()];
        int i = 0;
        for (Integer num : res) {
            result[i++] = num;
        }
        return result;
    }
}

File: MonotonicStack/496_Next_Greater_Element_I/496_Next_Greater_Element_I.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        
        stack<int> mystack;
        mystack.push(INT_MAX);
        unordered_map<int,int> hash;
        vector<int> res;

        for( const auto& num : nums2 ) {
            while(num > mystack.top()) {
                auto key = mystack.top();
                mystack.pop();
                hash.insert(make_pair(key, num)); 
            }
            mystack.push(num);        
        }

        for( const auto& num : nums1 ) {
            unordered_map<int, int>::const_iterator got = hash.find(num);
            if ( got == hash.end() ) {
                res.push_back(-1);
            } else {
                res.push_back(got->second);  
            }
        }

        return res;
    }
};
